File: src/extlib.mjs
import {
	Error,
	Ok,
} from "./gleam.mjs";

const Nil = undefined;

export function unix_timestamp() {
	return Math.floor(new Date().getTime() / 1000);
};

export function operating_system() {
	if (typeof navigator !== 'undefined') {
		return navigator.userAgent;
	}
	return `${operating_system_name()} ${operating_system_version()}`;
};

export function array_new() {
	let arr = [];
	return arr;
};

export function array_from_list(list) {
	let arr = list.toArray(); // Gleam thinks Array is a Tuple, on JS
	return arr;
};

export function array_to_list(arr) {
	return arr;
};

export function array_set(arr, idx, val) {
	arr[idx] = val;
	return arr;
};

export function array_get(arr, idx) {
	return idx in arr ? new Ok(arr[idx]) : new Error(Nil);
};

function operating_system_name() {
	const os = require('os');
	return os.platform();
};

function operating_system_version() {
	const os = require('os');
	return os.release();
};

File: src/extlib.erl
-module(extlib).

-export([unix_timestamp/0, operating_system/0, array_from_list/1, array_to_list/1,
         array_set/3, array_get/2, array_size/1, array_count/1]).

unix_timestamp() ->
    {MegaSecs, Secs, _} = os:timestamp(),
    MegaSecs * 1000000 + Secs.

operating_system() ->
    Name = operating_system_name(),
    Version = operating_system_version(),
    <<Name/binary, " ", Version/binary>>.

operating_system_name() ->
    {_OsFamilyAtom, OsNameAtom} = os:type(),
    OsNameString = atom_to_list(OsNameAtom),
    list_to_binary(OsNameString).

operating_system_version() ->
    _ = os:type(),
    OsVersion =
        case os:version() of
            {Major, Minor, Release} ->
                [integer_to_list(Major),
                 ".",
                 integer_to_list(Minor),
                 ".",
                 integer_to_list(Release)];
            VersionString ->
                VersionString
        end,
    list_to_binary(OsVersion).

array_from_list(List) ->
    array:from_list(List).

array_to_list(Array) ->
    array:sparse_to_list(Array).

array_set(Array, Index, Value) ->
    array:set(Index, Value, Array).

array_get(Array, Index) ->
    try array:get(Index, Array) of
        undefined ->
            {error, nil};
        Any ->
            {ok, Any}
    catch
        error:badarg ->
            {error, nil}
    end.

array_size(Array) ->
	array:sparse_size(Array).

array_count(Array) ->
	length(array:sparse_to_list(Array)).
